src/cargo/core/compiler/job_queue.rs

```diff
@@ -276,76 +278,90 @@ impl<'a> JobQueue<'a> {
             tokens.truncate(self.active.len() - 1);
 
-            let count = total - self.queue.len();
-            let active_names = self
-                .active
-                .iter()
-                .map(Key::name_for_progress)
-                .collect::<Vec<_>>();
-            drop(progress.tick_now(count, total, &format!(": {}", active_names.join(", "))));
-            let event = self.rx.recv().unwrap();
-            progress.clear();
-
-            match event {
-                Message::Run(cmd) => {
-                    cx.bcx
-                        .config
-                        .shell()
-                        .verbose(|c| c.status("Running", &cmd))?;
-                }
-                Message::BuildPlanMsg(module_name, cmd, filenames) => {
-                    plan.update(&module_name, &cmd, &filenames)?;
-                }
-                Message::Stdout(out) => {
-                    println!("{}", out);
-                }
-                Message::Stderr(err) => {
-                    let mut shell = cx.bcx.config.shell();
-                    shell.print_ansi(err.as_bytes())?;
-                    shell.err().write_all(b"\n")?;
-                }
-                Message::FixDiagnostic(msg) => {
-                    print.print(&msg)?;
-                }
-                Message::Finish(key, result) => {
-                    info!("end: {:?}", key);
-
-                    // self.active.remove_item(&key); // <- switch to this when stabilized.
-                    let pos = self
-                        .active
-                        .iter()
-                        .position(|k| *k == key)
-                        .expect("an unrecorded package has finished compiling");
-                    self.active.remove(pos);
-                    if !self.active.is_empty() {
-                        assert!(!tokens.is_empty());
-                        drop(tokens.pop());
+            // Drain all events at once to avoid displaying the progress bar
+            // unnecessarily.
+            let events: Vec<_> = self.rx.try_iter().collect();
+            let events = if events.is_empty() {
+                self.show_progress(total);
+                vec![self.rx.recv().unwrap()]
+            } else {
+                events
+            };
+
+            for event in events {
+                // CAUTION: Care must be taken to clear the progress bar if a
+                // message is to be actually displayed. Try to avoid
+                // unnecessarily clearing it to avoid flickering.
+                match event {
+                    Message::Run(cmd) => {
+                        if cx.bcx.config.shell().verbosity() == Verbosity::Verbose {
+                            self.progress.clear();
+                        }
+                        cx.bcx
+                            .config
+                            .shell()
+                            .verbose(|c| c.status("Running", &cmd))?;
```

Comment:
Since the verbosity check is already happening above, could this `c.status(...)` move into the `if` above?

Reply:
Similar with `"Fresh"` below